### DISCUSSION TITLE
Writing a base dialog component using dialog plugin "Invoking custom component"

### DISCUSSION THREAD

**Discussion by skycamdk** [quasarframework/quasar#10855] [General - Components / Directives / etc] [answered]
Hi,

I need several invokable dialogs. Following the examples from https://quasar.dev/quasar-plugins/dialog#invoking-custom-component works like a charm.

However I would like to avoid repeating myself in the components. Meaning not to have the base q-card design repeated in all my dialogs. I have been playing with making a "BaseDialog" component which contains the q-dialog, q-card layout and a slot for the parent to put the dialog specific components into.
Now I should be able to use that component as a base when writing dialogs. For an example:
A "ProfileEditDialog" that contains the BaseDialog component and where the dialog specifics are put in the BaseDialog Slot.

Now when trying to invoke the dialog I get a "Incorrectly defined Dialog component" error in the console. I found that it's because that my ProfileEditDialog does not have the show method or any of the other properties from the BaseDialog. So it seems that I need a way to expose all the properties from BaseDialog to the ProfileEditDialog.

Can someone explain how to expose a child's properties on the parent? Or some other way of getting q-dialog to work as I need.

Many thanks in advance!

**Comment by hawkeye64** [comment]
I do something like this at the end of `setup()`

```js
return () => h(component.value, { ref: calendar, ...props, ...attrs }, slots)
```
Where `component` is a computed value of what component to use based on property value passed in. I put a ref on it because, for each method, I need to duplicate it and pass it onto the parent.

Hopefully, this gives you some ideas.

**Reply by skycamdk** [reply]
Hi @hawkeye64,
Thank you for taking the time to help me. It's highly appreciated!

Unfortunately i'm not skilled enough to extract a working solution from what you described. Could you share code where you use this magic?

Thank you!

**Reply by hawkeye64** [reply]
https://github.com/quasarframework/quasar-ui-qcalendar/blob/next/ui/src/components/QCalendar.js#L130

**Comment by metalsadman** [comment, marked as answer]
the error log `"Incorrectly defined Dialog component"` means you didn't implement the ***required*** methods as you see in the docs. When you are using the re-usable component beside the plugin, then using a template ref like Jeff sited above is what you use to invoke inner methods/properties from the child, as a plugin there is no `show` function exposed so recreate the dialog. example:
https://codepen.io/metalsadman/pen/LYLqyrQ?editors=1010

**Comment by skycamdk** [comment]
Thank you both @hawkeye64 & @metalsadman,

You both sent me i the right direction. I think that my solution is an anti-pattern, however it's working.
 I did as you described @metalsadman and forwarded the "show" method from child to parent. Then "useDialogPluginComponent" accepted my dialog.

Next problem was to get the underlying BaseDialog (q-dialog) to append the parents data to its payload when firing "onOK" callback.
Form data lives in the "ProfileEditDialog" however it's the "BaseDialog" which fires the "onOK".
I ended up with using slot props so that ProfileEditDialogs formdata is referring to data inside BaseDialog scope.
It's working it is probably anti-pattern. Don't know...
